[CR]Fix crash when refusing skill reward

Coolthulhu opened the pull request:
Picking the option "teach me a skill" would call `talk_function::clear_mission`, but leave "TALK_MISSION_SUCCESS" on the stack before the player would select the reward.

As a result picking "TALK_NONE" would go back to topic being mission success (and reward selection), but now the mission would be a `nullptr`. And "TALK_MISSION_SUCCESS" unconditionally calls `miss->get_value()`.

I moved the `SUCCESS_ACTION(&talk_function::clear_mission);` from reward selection down to specific skill/style selection.

[CR] because I'm not sure if I did it right (ie. if the macro is valid where I used it). It did work in my tests, but might possibly not work in some cases.

I also did some minor refactoring and added a DEBUG_MIND_CONTROL mutation that allows talking to hostile NPCs, demanding that they become friendly and makes all attempts at convincing NPCs successful.

BevapDin commented on src/game.cpp:
```diff
@@ -11584,29 +11583,33 @@ void game::chat()
-        u.shout();
-
-    } else if (available.size() == 1) {
+    if( available.size() == 1 ) {
```
So, if there is exactly *one* NPC nearby, I will automatically talk to them and won't have the chance to yell. If there are several or none nearby, I can shout. Seems strange.

Btw. I see some code duplication: `talk_to_u(); moves -= 100; refresh_all();` is called here, and below after choosing an NPC. This should be in a function, or only at one place. If I wanted to change the moves based on what `talk_to_u` returns (the time of the talking), I would have to remember to change it at two separate places.

Coolthulhu replied:
As for the `available.size() == 1` - it was like this before and currently there is "Y - yell" in NPC menu. The only mechanics change here is the `available.empty()` case - currently (in master) it has a separate y/n prompt, here it allows yell/cancel (for muscle memory and consistency).

I could move it all to this menu and remove the "exactly one" case.

BevapDin replied:
> As for the  available.size() == 1  - it was like this before and currently there is "Y - yell" in NPC menu. The only mechanics change here is the  available.empty()  case - currently (in master) it has a separate y/n prompt, here it allows yell/cancel (for muscle memory and consistency).

Is there any other (standard) way to shout? If so, it's probably better to keep your code: if there is exactly one NPC nearby, talk to that one (after all, the player pressed 'C' to chat, not to shout), otherwise show the menu with the additional shout option.

If there is no other way to shout, always opening the menu is probably better as it allows the player to press a fixed key sequence, 'C' (chat, brings up the menu) and 'a' (shout, always an option), to shout, regardless of any nearby NPCs.

BevapDin commented:
Now (or maybe this is an old bug?) it crashe swhen I answer "On second thought, never mind." after the NPC asks TALK_TRAIN_START.

I think there is another potential problem here: TALK_TRAIN can be reached without any mission, just asking follower to train you will reach it. I think a solution could be to add an effect that states the NPC owns you a training session, which allows bypassing the "asked_to_train" and probably bypassing the cost for training.

But I very much like the mind control trait, why is it not a normal mutation? (-;

Coolthulhu commented:
Changed it so that the mission is only cleared when the training actually starts. Since `start_training` is only called as a reward, it shouldn't cause problems.

Also added checks for mission not being null. If it manages to be, it will print a debugmsg instead of crashing.
